### PR TITLE
Feature/65 oauth server

### LINKED ISSUE #65: Enable the CAS server to act as an OAuth server

### PR DESCRIPTION
**Pull request by Subaro**
Resolves #65

### CODE REVIEW

File: app/src/main/java/de/triology/cas/services/RegistryEtcd.java [hunk 1]
```diff
@@ -199,4 +246,89 @@ protected JSONObject getCurrentDoguNode(String doguName) throws ParseException {
...
+
+    /**
+     * Retrieves the client secret for a given clientId and encrypts the secret accordingly.
+     *
+     * @param clientID Identifier for the OAuth client
+     * @return hash of the actual client secret
+     * <p>
+     */
+    protected String getCurrentOAuthClientSecret(String clientID) {
+
```
Comment (Subaro) [inline app/src/main/java/de/triology/cas/services/RegistryEtcd.java]: We currently only save the hash of the password non decrypted in the etcd. This is open for discussion.

File: app/src/main/java/de/triology/cas/services/CesServicesManagerStageProductive.java [hunk 2]
```diff
@@ -92,15 +107,22 @@ private void registerChangeListener() {
...
+                addNewService(service);
+            } catch (GetCasLogoutUriException e) {
+                logger.info("GetCasLogoutUriException: CAS logout URI of service " + serviceData.getName() + " could not be retrieved: " + e);
+                logger.info("Adding service without CAS logout URI");
+                RegexRegisteredService service = serviceData.getFactory().createNewService(createId(), fqdn, null, serviceData);
+                addNewService(service);
+            }
+        }
+        catch (Exception e) {
+            logger.info("Could not create service " + serviceData.getName() + ": " + e);
```
Comment (sklein94) [inline app/src/main/java/de/triology/cas/services/CesServicesManagerStageProductive.java:125]: Whi is this info and not error?

File: app/src/main/java/de/triology/cas/services/RegistryEtcd.java [hunk 3]
```diff
@@ -18,19 +19,22 @@
...
 import java.net.URISyntaxException;
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
 import java.util.concurrent.TimeoutException;
 
 /**
  * Default implementation of {@link Registry} using {@link EtcdClient}.
- *
+ * <p>
```
Comment (sklein94) [inline app/src/main/java/de/triology/cas/services/RegistryEtcd.java:28]: why?

File: app/src/main/java/de/triology/cas/services/RegistryEtcd.java [hunk 4]
```diff
@@ -44,12 +48,74 @@ public RegistryEtcd(EtcdClient etcd) {
...
+        for (EtcdKeysResponse.EtcdNode oAuthClient : nodesFromEtcd) {
+            try {
+                String clientID = oAuthClient.getKey().substring(CAS_SERVICE_ACCOUNT_DIR.length());
+                String clientSecret = this.getCurrentOAuthClientSecret(clientID);
+                HashMap<String, String> attributes = new HashMap<>();
+                attributes.put(CesOAuthServiceFactory.ATTRIBUTE_KEY_OAUTH_CLIENT_ID, clientID);
+                attributes.put(CesOAuthServiceFactory.ATTRIBUTE_KEY_OAUTH_CLIENT_SECRET, clientSecret);
+                serviceAccountNames.add(new CesServiceData(clientID, factory, attributes));
+            } catch (RegistryException ex) {
+                log.warn("registry exception occurred", ex);
```
Comment (sklein94) [inline app/src/main/java/de/triology/cas/services/RegistryEtcd.java:87]: error

File: app/src/main/java/de/triology/cas/services/RegistryEtcd.java [hunk 5]
```diff
@@ -44,12 +48,74 @@ public RegistryEtcd(EtcdClient etcd) {
...
+        for (EtcdKeysResponse.EtcdNode dogu : nodesFromEtcd) {
+            JSONObject json;
+            try {
+                String doguName = dogu.getKey().substring(DOGU_DIR.length());
+                json = getCurrentDoguNode(doguName);
+                if (hasCasDependency(json)) {
+                    doguServices.add(new CesServiceData(doguName, factory));
+                }
+            } catch (ParseException ex) {
+                log.warn("failed to parse EtcdNode to json", ex);
```
Comment (sklein94) [inline app/src/main/java/de/triology/cas/services/RegistryEtcd.java:105]: error?

File: app/src/main/java/de/triology/cas/services/oauth/CesOAuth20CallbackAuthorizeController.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        model.put("callbackUrl", callbackUrl);
+
+        /*
+        // retrieve service name from session
+        final String serviceName = (String) session.getAttribute(OAuthConstants.OAUTH20_SERVICE_NAME);
+        logger.debug("serviceName : {}", serviceName);
+        model.put("serviceName", serviceName);
+
+        return new ModelAndView(OAuthConstants.CONFIRM_VIEW, model);
+        */
```
Comment (sklein94) [inline app/src/main/java/de/triology/cas/services/oauth/CesOAuth20CallbackAuthorizeController.java:86]: dead code

File: app/src/main/java/de/triology/cas/services/oauth/CesOAuth20CallbackAuthorizeController.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * @author Jerome Leleu
+ * @since 3.5.0
+ */
+public final class CesOAuth20CallbackAuthorizeController extends AbstractController {
+
+    private final Logger logger = LoggerFactory.getLogger(CesOAuth20CallbackAuthorizeController.class);
+
+    @Override
+    protected ModelAndView handleRequestInternal(final HttpServletRequest request, final HttpServletResponse response)
```
Comment (sklein94) [inline app/src/main/java/de/triology/cas/services/oauth/CesOAuth20CallbackAuthorizeController.java:48]: override?

File: app/src/main/java/de/triology/cas/services/oauth/CesOAuthServiceFactory.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+     * @param id Internal ID of the service
+     * @return a new OAuth callback authorisation service
+     */
+    public OAuthCallbackAuthorizeService createCallbackService(long id, String fqdn) {
+        OAuthCallbackAuthorizeService service = new OAuthCallbackAuthorizeService();
+        service.setId(id);
+        service.setName(CesOAuthServiceFactory.class.getSimpleName() + " " + SERVICE_OAUTH_CALLBACK_IDENTIFIER);
+        String serviceId = "https://" + fqdn + "(:443)?/" + "oauth2.0/callbackAuthorize";
+        service.setServiceId(serviceId);
+        service.setAllowedToProxy(true);
```
Comment (sklein94) [inline app/src/main/java/de/triology/cas/services/oauth/CesOAuthServiceFactory.java:33]: remove

File: app/src/main/webapp/WEB-INF/cas.properties [hunk 9]
```diff
@@ -205,3 +205,16 @@ login.limit.maxNumber=0
 login.limit.failureStoreTime=0
 login.limit.lockTime=0
 login.limit.maxAccounts=10000
+
+#========================================
+# OAuth Settings
```
Comment (sklein94) [inline app/src/main/webapp/WEB-INF/cas.properties:210]: remove file

**Review by sklein94 — CHANGES_REQUESTED** [review]

File: app/src/main/java/de/triology/cas/logout/LogoutUriEnabledLogoutManagerImpl.java [hunk 10]
```diff
@@ -128,6 +127,12 @@ public LogoutUriEnabledLogoutManagerImpl(final ServicesManager servicesManager,
         return logoutRequests;
     }
 
+    /**
+     *
+     * @param singleLogoutService
+     * @param logoutRequest
+     * @param registeredService
+     */
```
[the remark this answers is not in the document]
Comment (Subaro) [reply app/src/main/java/de/triology/cas/logout/LogoutUriEnabledLogoutManagerImpl.java:135]: Resolved

File: app/src/main/java/de/triology/cas/services/CesServicesManagerStageProductive.java (hunk 2 above)
Comment (Subaro) [reply app/src/main/java/de/triology/cas/services/CesServicesManagerStageProductive.java:125]: Should be error, fixed

File: app/src/main/java/de/triology/cas/services/RegistryEtcd.java (hunk 3 above)
Comment (Subaro) [reply app/src/main/java/de/triology/cas/services/RegistryEtcd.java:28]:
```
 * Default implementation of {@link Registry} using {@link EtcdClient}.
 * <p>
 * The Dogus are queried from etcd: Installed Dogus and the version information are stored in a directory
 * <code>/dogu/${name of dogu}/current</code>. In addition, 'cas' has to be in the dependencies of the Dogu.
 * Changes of the <code>/dogu</code> directory can be recognized using {@link #addDoguChangeListener(DoguChangeListener)}.
 ```

Comment (Subaro) [reply app/src/main/java/de/triology/cas/services/RegistryEtcd.java:28]: Seems reasonable to me

File: app/src/main/java/de/triology/cas/services/RegistryEtcd.java (hunk 4 above)
Comment (Subaro) [reply app/src/main/java/de/triology/cas/services/RegistryEtcd.java:87]: Yes

File: app/src/main/java/de/triology/cas/services/RegistryEtcd.java (hunk 5 above)
Comment (Subaro) [reply app/src/main/java/de/triology/cas/services/RegistryEtcd.java:105]: Yes

File: app/src/main/java/de/triology/cas/services/oauth/CesOAuth20CallbackAuthorizeController.java (hunk 6 above)
Comment (Subaro) [reply app/src/main/java/de/triology/cas/services/oauth/CesOAuth20CallbackAuthorizeController.java:86]: removed

File: app/src/main/java/de/triology/cas/services/oauth/CesOAuth20CallbackAuthorizeController.java (hunk 7 above)
Comment (Subaro) [reply app/src/main/java/de/triology/cas/services/oauth/CesOAuth20CallbackAuthorizeController.java:48]: no it brings no advantages

File: app/src/main/java/de/triology/cas/services/oauth/CesOAuthServiceFactory.java (hunk 8 above)
Comment (Subaro) [reply app/src/main/java/de/triology/cas/services/oauth/CesOAuthServiceFactory.java:33]: Removed

File: app/src/main/webapp/WEB-INF/cas.properties (hunk 9 above)
Comment (Subaro) [reply app/src/main/webapp/WEB-INF/cas.properties:210]: Removed

File: app/src/main/java/de/triology/cas/services/RegistryEtcd.java (hunk 1 above)
Comment (sklein94) [reply app/src/main/java/de/triology/cas/services/RegistryEtcd.java]: New ticket will be created.

**Review by sklein94 — COMMENTED** [review]
looks good

**Pull request merged**